tests/e2e/test_code_watcher.py

```diff
@@ -0,0 +1,134 @@
+import unittest
+import subprocess
+import os
+from pathlib import Path
+from time import sleep, time
+import threading
+
+def read_output(pipe, output_list):
+    """
+    Read lines from a subprocess pipe and append them to the output list.
+    """
+    for line in iter(pipe.readline, ''):
+        output_list.append(line.strip())
+
+class TestCodeWatcher(unittest.TestCase):
+    def setUp(self):
+        """
+        Set up the necessary environment for testing.
+        """
+        self.test_dir = Path("/usr/local/submitty/GIT_CHECKOUT/Submitty/")
+        self.setup_dir = Path(self.test_dir, ".setup")
+        self.site_dir = Path(self.test_dir, "site")
+        self.bin_dir = Path(self.test_dir, "bin")
+        self.sbin_dir = Path(self.test_dir, "sbin")
+
+        # Ensure directories exist
+        for directory in [self.setup_dir, self.site_dir, self.bin_dir, self.sbin_dir]:
+            if not directory.exists():
+                self.skipTest(f"Required directory {directory} does not exist.")
+
+        self.script_path = Path("/usr/local/submitty/GIT_CHECKOUT/Submitty/.setup/bin/code_watcher.py")
```

Comment:
This could be simplified by using some of the previously set variables to reduce possible points of confusion later if the path is changed.